### DISCUSSION TITLE
Adding code / functionality at runtime

### DISCUSSION THREAD

**Discussion by bcluap** [quarkusio/quarkus#33279] [Q&A]
Has anyone got any good ideas on the best way to add source code (e.g. a class implementing an interface) at runtime in Quarkus and then call that code? We have used javassist for this and it works fine but only supports java 8 source  so was wondering if there is a newer better way. The problem we are trying to solve is to allow custom functionality/ plugins to be loaded at runtime to augment an existing products business logic without restarting or redeploying a running saas.

**Comment by bcluap** [comment]
Any thoughts on this?

**Reply by pigelvy** [reply]
This is an interesting question! I am looking for a way to build a Quarkus application with its own API that would be enriched by plugins. Of course, the application cannot know about its plugins.

Ideally, the for a given project/use-case, the required plugins would be copied at install-time to some folder and "discovered" at runtime (whatever that means for Quarkus). The plugins would be able to things like:
 - Implement application API interfaces and declared them as injectable beans
 - Provide their own REST endpoints

Is that even possible in Quarkus? Maybe this way to doing things is wrong by the Quarkus philosophy?

As a last resort, another option I see to develop our extensible application would be to develop the core of the application as a quarkus-indexed library, and create a real Quarkus project for the each project/use-case. This way, the necessary plugins would be known at build-time and everything should be more quarkus-friendly.

Any help/advice?

**Reply by bcluap** [reply]
We have solved for the use case of runtime bean providers that can change implementation at runtime. This uses CDI.current().getBeanManager() and finds the implementation we want which can be bound at runtime and changed at runtime through config. Completely dynamic code though relies on JavaAssist at this point
